### PR TITLE
Feature/configure abstract

### PR DESCRIPTION
**Pull request by bmoore20**
Hey Ty! Can you please take a look at these changes when you get the chance and let me know what you think? Thanks! @AlephNotation 

This PR includes two main changes: 

1. Configuration of the hyperparameters (#37)
2. Abstraction of the training and testing loops (#38, #17 )

I combined them into one PR because their changes are dependent on each other. 

For the configuration of the hyperparameters, I created a new file under the `utils` folder called `selectors`. For the hyperparameters that were objects (model, criterion, and optimizer) I needed a way to initiate the specific type of objects that the user passed in before I gave them to the `train` method call. Pulling this functionality out into a separate file was the cleanest option. All of the selector calls are called in `main` in _train.py_. Currently, each selector only has one option, so it seems like a little bit of an overkill. But I am anticipating that once we start experimenting with the model, we will want to add more options to these selectors. With this design in will be easy to add in new types of models, optimizers, and criterions.

For the abstraction of the training and test loops, I created a new file under the `utils` folder called `training_helper`. This file consists of two methods, `training_laps` and `evaluate`. `training_laps` consists of the functionality that trains the dataset, while `evaluate` pulls out the code that handles the evaluation of the trained model. Both are called in `train` in _train.py_. By pulling out the training and testing loops, _train.py_ now has two main jobs: 1) load in the data 2) call the train and test functions. Overall, _train.py_ is now a lot easier to read and follow.

### CODE REVIEW

File: hab/utils/training_helper.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    """
+    Execute training on model by completing specified number of epochs/laps through database.
+
+    :param model:Model to be trained.
+    :param data_loader: Data loader that contains training data.
+    :param epochs: Number of laps to complete through dataset.
+    :param optimizer: Optimization algorithm used to train the model.
+    :param criterion: Loss function used to train the model.
+    """
+    for epoch in range(epochs):
```
Comment (AlephNotation) [inline hab/utils/training_helper.py:39]: Since you call `model.eval()` in the eval function, you need to call `model.train()` before your training loop starts. Otherwise none of the regularization will be present during training.

File: hab/train.py [hunk 2]
```diff
@@ -62,60 +65,37 @@ def train(train_data_dir: str, test_data_dir: str, magnitude_increase: int = 1):
...
-            _, predicted = torch.max(outputs.data, 1)
-            total += targets.size(0)
-            correct += (predicted == targets).sum().item()
-
-    logger.info("Accuracy of the network on the 10000 test images: %d %%" % (
-            100 * correct / total))
-
-
-def main(train_dataset: str, test_dataset: str):
+    evaluate(model, test_loader)
```
Comment (AlephNotation) [inline hab/train.py:74]: If this is *not* a holdout set, you want to move this thing into the training loop after the batchloader for loops. Think of it sort of like quizzes. After every iteration through the material, you want to quiz your model and see how things are going. Plus this will also let you introduce learn rate optimizers.

File: hab/utils/selectors.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+def optimizer_selector(optim_type: str) -> Optimizer:
+    """
+    Retrieve specified optimizer object.
+
+    :param optim_type: Name of desired optimizer.
+    :return: Instance of Optimizer object.
+    """
+    if optim_type == "Adam":
+        return optim.Adam(lr=1e-3)
```
Comment (AlephNotation) [inline hab/utils/selectors.py:43]: This should return an uninstantiated `torch.nn.optimizer` object. The learn rate is one of the main hyperparams you'll need to tune.

File: hab/utils/training_helper.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        for data in data_loader:
+            images, targets = data
+
+            outputs = model(images)  # nn.module __call__()
+            _, predicted = torch.max(outputs.data, 1)
+            total += targets.size(0)
+            correct += (predicted == targets).sum().item()
+
+    logger.info("Accuracy of the network on the 10000 test images: %d %%" % (
+            100 * correct / total))
```
Comment (AlephNotation) [inline hab/utils/training_helper.py:81]: need a blank line to end the file :)

File: hab/utils/selectors.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    else:
+        raise ValueError(f"Optimizer type must be Adam. Value received: {optim_type}")
```
Comment (AlephNotation) [inline hab/utils/selectors.py:45]: blank line to end a the file

**Review by AlephNotation — CHANGES_REQUESTED** [review]
Overall excellent work! Let's discuss the training lap construction during our call.

As a side note, have you thought about using a code formatter? I'm personally a huge fan of black.

File: hab/utils/selectors.py (hunk 3 above)
Comment (bmoore20) [reply hab/utils/selectors.py]: I googled and looked for a `torch.nn.optimizer` object, but I had no luck. The PyTorch documentation kept referring me to the `torch.optim.Optimizer` interface.  

https://pytorch.org/docs/master/optim.html#base-class

I pulled out and configured the learn rate for the `torch.optim.Opimizer` objects similar to what I did for the other hyperparameters.

File: hab/utils/training_helper.py (hunk 4 above)
Comment (bmoore20) [reply hab/utils/training_helper.py]: added a blank line but it's not showing up for some reason

File: hab/utils/selectors.py (hunk 5 above)
Comment (bmoore20) [reply hab/utils/selectors.py:47]: added a blank line but it's not showing up for some reason

File: hab/utils/training_helper.py (hunk 1 above)
Comment (bmoore20) [reply hab/utils/training_helper.py]: Ok thanks. I called `model.eval()` in my new `validation_lap` function and `model.train()` in the `training_lap` function. Is that what you were talking about?

File: hab/train.py (hunk 2 above)
Comment (bmoore20) [reply hab/train.py:82]: My original thinking was that the `evaluate` method would consist of a holdout set that would test on the final trained model. In order to test the model as it trains like you mentioned during our call, I created a 3rd dataset, `validation dataset`. This also contains a holdout set, but it is now given after the `training_lap` inside of the training loop. Is adding a `validation lap` what you were talking about here?

File: hab/utils/training_helper.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            images, targets = data
+
+            optimizer.zero_grad()
+
+            outputs = model(images)  # nn.module __call__()
+            loss = criterion(outputs, targets)
+            loss.backward()
+            optimizer.step()
+
+            running_loss += loss.item()
```
[the remark this answers is not in the document]
Comment (bmoore20) [reply hab/utils/training_helper.py]: Good idea! I pulled them out into their own functions. They can be found in `training_helper.py`. I had to make changes in _HABsDataset.py_ and _train.py_ in order to support the new **validation dataset** needed for the `validation_lap`.

I still have to figure out exactly what to log for the losses training loop. It gets a little weird with scope of variables.

**Comment by bmoore20** [comment]
Thanks for the great feedback @AlephNotation !

Also - thanks for the reminder about using Black. I just set it up with the repo using pre-commits so it should be all set now! I submitted and merged the PR (#41) for it so it's in `main` already.

**Comment by bmoore20** [comment]
@AlephNotation I am going to merge the code in this PR because we went through all of the original changes that I made during our call yesterday and you said everything looked good. The commits I made today were all little changes that you suggested yesterday. A list of these changes are:

- In training lap `for epoch in range(epoch):` log the **epoch #** and the **training loss** to a 7-point decimal
- Save the model 
- Give batch size to `DataLoader` object and do `for batch in data_loader:` in _training_helper.py_
- Configure batch parameter
- Fix current model architecture to handle **3 classifications** 

Time to **_FINALLY_** run the program!! Thanks for your help!

File: hab/utils/training_helper.py (hunk 1 above)
Comment (bmoore20) [reply hab/utils/training_helper.py]: During our call on 5.5.21 Ty confirmed this was correct.

File: hab/train.py (hunk 2 above)
Comment (bmoore20) [reply hab/train.py:101]: During our call on 5.5.21 Ty confirmed this was correct.
